Checkout-payment branch merge

sara-db opened the pull request:
~From branch cart~
~Folder restructure~
~Introduction of notion of sessions~
~Cart page added~ 
~“Add to cart” CTA on product page functional~ 
~“Remove item”~ 
~“Update quantity”~ 
~Cart and CartController Refactoring to remove duplication~

~From branch checkout-customer-information~:
~Checkout now CTA on cart page links to checkout page~ 
~Checkout page structure added~ 
~Validate and submit form~
~Redirect to payment page~

_Comment edited by PK because each branch was merged individually_

From branch checkout-payment:
- Payment page and thank you page added
- Payment form validation and submission
- Schema update
- Order data saved on database

sara-db committed:
Payment page added with form validation and submission

sara-db committed:
Merge branch 'checkout-customer-information' into checkout-payment

sara-db committed:
Data saved to database after payment form is submitted. Thank you page added.

sara-db committed:
Synced with master. Conflitcts resolved.

paulkennethkent committed:
Resolved merge conflicts

* master:
  update copyright to 2018
  Improve README.md
  Removed unused code Removed unused code from default.html

paulkennethkent commented:
@JohnBergqvist, @Charliemowood and @jgwilson42 

Please could you review this PR. 

This is the last of @sara-db mega PR that has been split and tested. I then have one last PR for the homepage and then the tax code changes.

Charliemowood commented:
![screen shot 2018-02-12 at 15 50 23](https://user-images.githubusercontent.com/25271307/36105447-8498bd60-100c-11e8-9796-493e7d49125d.png)

Probably some guards should be introduced to force the customer to put in the correct data.

jgwilson42 commented on src/main/java/com/diffblue/demo/ecommerce/controllers/CheckoutController.java:
```diff
@@ -87,4 +109,78 @@ public String viewPaymentPage(Map<String, Object> model, HttpSession session) {
+   * @return page for the output
+   */
+  @PostMapping("/payment")
+  public String makePayment(@Valid PaymentForm paymentForm, BindingResult bindingResult,
```
Please, could you add comments to the method? It's almost worth over commenting based on us using this as a demo.

jgwilson42 commented on src/main/java/com/diffblue/demo/ecommerce/models/OrderItem.java:
```diff
@@ -0,0 +1,81 @@
+package com.diffblue.demo.ecommerce.models;
+
+// Copyright 2016-2017 DiffBlue limited. All rights reserved.
```
Please could you update the copyright notices to 2016-2018

jgwilson42 commented on src/main/java/com/diffblue/demo/ecommerce/repositories/ProductRepository.java:
```diff
@@ -5,6 +5,7 @@
 import com.diffblue.demo.ecommerce.models.Product;
 import org.springframework.data.repository.Repository;
 
+
```
Is this the only change to the file in this PR? If so, please remove.

jgwilson42 commented on src/main/resources/db/hsqldb/schema.sql:
```diff
@@ -20,14 +28,52 @@ DROP TABLE products IF EXISTS;
-    photo VARCHAR(500),
-    FOREIGN KEY (category_id) REFERENCES categories(id),
-    FOREIGN KEY (collection_id) REFERENCES collections(id)
+  id              INTEGER IDENTITY PRIMARY KEY,
```
Is the indenting correct here?

paulkennethkent replied:
@jgwilson42 I've seen that style in a few examples but I've also seen no indenting. 

I can change it if you want?

JohnLBergqvist replied:
I would also prefer reducing the indenting tbh.

jgwilson42 commented on src/main/resources/templates/OrderComplete.html:
```diff
@@ -0,0 +1,28 @@
+            <div class="row">
+                <div class="col-12">
+                    <p><b>Questions?</b></p>
+                    <p><b>Call 01865 123 456 <br/> or <br/> <a href="mailto:test@test.com">Email us</a></b></p>
```
I would make these Diffblue contact details, e.g. Diffblue phone number and test@diffblue.com

This prevents anyone seeing this and calling/e-mailing  (unlikely I know).

jgwilson42 reviewed (changes requested):
Hi Paul,

A few changes but otherwise looks good.

Thanks,
James

JohnLBergqvist reviewed (commented):
Personally, I don't like the indenting style, both on the SQL statements and on some of the function parameters (like the `Order` constructor). I'm no code style expert though, so maybe you're supposed to do it like that for function parameters.

jgwilson42 commented on src/main/java/com/diffblue/demo/ecommerce/forms/AddressForm.java:
```diff
@@ -20,7 +20,6 @@
   @NotEmpty
   private String postcode;
 
-
```
Does this file only have whitespace changes? It would make sense to remove the change from the PR.

jgwilson42 reviewed (approved):
Hi Paul,

Just one suggestion, otherwise looking good.

Thanks,
James

paulkennethkent committed:
Added html 5 form validation to stop errors
Validaiton added to checkout and payment key fields

Fixes to validation after testing
There were some error with the previous commit, they will be squashed eventually

Changes to code styling, copyright dates and form validation
A few code styling changes to contructores and the schema sql. The copyright has been modified for this PR files and we brought in some HTML.

Code style changes
Reverting bak to old code styles for contructors

Fixed for the final time
Problems with validation again but I have refined the REGEX

removed whitespace
